### PR TITLE
autodoc: Respect type annotations in `__new__`, `metacls.__call__`, and builtin types

### PR DESCRIPTION
**Pull request by eric-wieser**
### Feature or Bugfix

- Bugfix

### Purpose
None of the following work correctly before this patch
* Signatures defined by `__new__`
* Signatures defined by metaclasses
* Signatures defined by builtin base classes

Patch includes a test showing that these now all work.

It's possible other tests will have been affected, will update if needed.

Ping @tk0miya, since you recently looked at annotation stuff.

### CODE REVIEW

**Comment by eric-wieser** [comment]
Failures are all of the form
```
E         -  '.. py:class:: DocstringSig()',
E         ?                             --
E         +  '.. py:class:: DocstringSig',
```

I think this change is correct - with it, these two classes are now documented identically:
```
class WithInit:
    def __init__(self):
        pass
class WithoutInit:
    pass
```
Whereas previously the former would have `()` but the latter would not.

The two classes give exactly the same result from `inspect.signature`, so I would consider this a bug.
I've added a second commit to fix this.

File: sphinx/ext/autodoc/__init__.py [hunk 1]
```diff
@@ -1175,22 +1175,73 @@ def format_args(self, **kwargs: Any) -> str:
...
+            call = get_user_defined(type(self.object), '__call__')
+
+            if call is not None:
+                # types which have confusing metaclass signatures it would be best
+                # not to show
+                call_blacklist = [
+                    'enum.EnumMeta.__call__',
+                ]
+                if "{0.__module__}.{0.__qualname__}".format(call) in call_blacklist:
+                    call = None
```
Comment (eric-wieser) [inline sphinx/ext/autodoc/__init__.py]: This is a little unfortunate, because for an enum
```python
class MyEnum(enum.Enum):
    VAL = 1
```
the signature is actually:
```python
>>> inspect.signature(MyEnum)
<Signature (value, names=None, *, module=None, qualname=None, type=None, start=1)>
```
This signature isn't actually a lie, it's just something no sane person would ever call:
```python
>>> MyEnum("MyTotallyNewEnum", names="why did i do this")
<enum MyTotallyNewEnum>
```

File: sphinx/ext/autodoc/__init__.py [hunk 2]
```diff
@@ -1175,22 +1175,73 @@ def format_args(self, **kwargs: Any) -> str:
...
-        # classes without __init__ method, default __init__ or
-        # __init__ written in C?
-        if initmeth is None or \
-                inspect.is_builtin_class_method(self.object, '__init__') or \
-                not(inspect.ismethod(initmeth) or inspect.isfunction(initmeth)):
-            return None
-        try:
-            self.env.app.emit('autodoc-before-process-signature', initmeth, True)
-            sig = inspect.signature(initmeth, bound_method=True)
+        def get_user_defined(obj, attr):
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py]: Could you add annotations please? And it would be better to add docstring if possible.

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: nits: `get_user_defined_function()` or `get_user_defined_function_or_method()` is better to me. What do you think?

File: sphinx/ext/autodoc/__init__.py [hunk 3]
```diff
@@ -1175,22 +1175,73 @@ def format_args(self, **kwargs: Any) -> str:
...
+            if inspect.is_builtin_class_method(obj, attr):
+                return None
+            attr = self.get_attr(obj, attr, None)
+            if not (inspect.ismethod(attr) or inspect.isfunction(attr)):
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py:1262]: FWIW, some kind of cython functions are not treated as "function" (refs: #7331). I'm not good at cython, but it might be needed to check `inspect.isroutine()`.

File: sphinx/ext/autodoc/__init__.py [hunk 4]
```diff
@@ -1175,22 +1175,73 @@ def format_args(self, **kwargs: Any) -> str:
...
+        # in its metaclass
+        if sig is None:
+            call = get_user_defined(type(self.object), '__call__')
+
+            if call is not None:
+                # types which have confusing metaclass signatures it would be best
+                # not to show
+                call_blacklist = [
+                    'enum.EnumMeta.__call__',
+                ]
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py]: I think such constants should be placed at the module level and be capitalized.

File: sphinx/ext/autodoc/__init__.py (hunk 1 above)
Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: Reasonable.

File: sphinx/ext/autodoc/__init__.py [hunk 5]
```diff
@@ -1175,22 +1175,73 @@ def format_args(self, **kwargs: Any) -> str:
...
+                if "{0.__module__}.{0.__qualname__}".format(call) in call_blacklist:
+                    call = None
+
+            if call is not None:
+                self.env.app.emit('autodoc-before-process-signature', call, True)
+                try:
+                    sig = inspect.signature(call, bound_method=True)
+                except TypeError:
+                    print("Faild, oh no")
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py]: This must be updated. Should this be replaced with a warning? Or be suppressed?

File: sphinx/ext/autodoc/__init__.py [hunk 6]
```diff
@@ -1175,22 +1175,73 @@ def format_args(self, **kwargs: Any) -> str:
...
+            init = get_user_defined(self.object, '__init__')
+            if init is not None:
+                self.env.app.emit('autodoc-before-process-signature', init, True)
+                try:
+                    sig = inspect.signature(init, bound_method=True)
+                except TypeError:
+                    print("Faild, oh no")
+
+        # if none of the attributes are user-defined, we can skip
+        # `autodoc-before-process-signature`, and just let inspect handle it
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py]: Why skipped here?

File: tests/test_ext_autodoc_configs.py [hunk 7]
```diff
@@ -508,8 +521,9 @@ def test_autodoc_typehints_signature(app):
         '',
         '.. py:function:: missing_attr(c, a: str, b: Optional[str] = None) -> str',
         '   :module: target.typehints',
-        ''
+        '',
     ]
+    assert list(actual) == expected
```
Comment (tk0miya) [inline tests/test_ext_autodoc_configs.py]: +0 for changing style. But all other cases should be updated also. Anyway, that should be another PR.

File: tests/test_autodoc.py [hunk 8]
```diff
@@ -510,14 +545,14 @@ def test_autodoc_members(app):
     # default (no-members)
     actual = do_autodoc(app, 'class', 'target.inheritance.Base')
     assert list(filter(lambda l: '::' in l, actual)) == [
-        '.. py:class:: Base',
+        '.. py:class:: Base()',
```
Comment (tk0miya) [inline tests/test_autodoc.py]: Note: `py:class` directive skips empty parenthesis. So this change harms nobody.

**Comment by tk0miya** [comment]
Should we also modify `get_doc()` to obtain docstring in order of this sequence?

File: sphinx/ext/autodoc/__init__.py (hunk 3 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py:1262]: This is the same logic as what was there before, so I'd like to declare this out of scope.

File: sphinx/ext/autodoc/__init__.py (hunk 5 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: Whoops, this was a debug statement!

File: sphinx/ext/autodoc/__init__.py (hunk 6 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: Because if the type is native, there's no source code to process, and the assumption is `autodoc-before-process-signature` has nothing to do.

File: tests/test_ext_autodoc_configs.py (hunk 7 above)
Comment (eric-wieser) [reply tests/test_ext_autodoc_configs.py]: This was leftover from a previous change which needed the separate variable. Will revert.

File: tests/test_autodoc.py (hunk 8 above)
Comment (eric-wieser) [reply tests/test_autodoc.py]: Excellent news.

**Comment by eric-wieser** [comment]
> Should we also modify get_doc() to obtain docstring in order of this sequence?

We could, but I don't think that matters so much, and would prefer not to do it in this PR.

File: sphinx/ext/autodoc/__init__.py (hunk 5 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: Be suppressed, because that's what this code did before.

File: sphinx/ext/autodoc/__init__.py (hunk 2 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: Updated to use `get_user_defined_function_or_method`, no opinion from me on the name.
Annotation and docstring added.

File: sphinx/ext/autodoc/__init__.py (hunk 6 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: Updated the comment a little.

**Comment by eric-wieser** [comment]
Comments addressed I think

File: tests/test_util_inspect.py [hunk 9]
```diff
@@ -29,12 +29,14 @@ def test_signature():
...
 
-    # builitin classes
-    with pytest.raises(TypeError):
-        inspect.signature(int)
-
-    with pytest.raises(TypeError):
-        inspect.signature(str)
+    # builtins are supported on a case-by-case basis, depending on whether
+    # they define __text_signature__
+    if getattr(list, '__text_signature__', None):
```
Comment (tk0miya) [inline tests/test_util_inspect.py:34]: I think this if-condition will be removed in the future, right?
It seems python3.7+ supports inspection for the list object. How about change this to `sys.version_info > (3, 7)`?

File: sphinx/ext/autodoc/__init__.py (hunk 6 above)
Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: I think this event is public one. So it should be emitted even if no source code.

**Review by tk0miya — CHANGES_REQUESTED** [review]
Sorry for late. Let's go forward this.

File: tests/test_util_inspect.py (hunk 9 above)
Comment (eric-wieser) [reply tests/test_util_inspect.py:34]: Could do, although I wonder if running python with `-OO` affects whether whether these attributes are populated.

At any rate, the presence of documentation on builtins sort of feels like an implementation detail. Switching to a version check is likely to break pypy or something.

File: sphinx/ext/autodoc/__init__.py (hunk 6 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: The only way to ensure it is emitted would be to either:
* Emit it for the object itself on this line, rather than attempting to find the right `__new__` / `__init__` etc
* Duplicating even more of the logic inside `inspect.signature`

File: sphinx/ext/autodoc/__init__.py [hunk 10]
```diff
@@ -1175,22 +1183,70 @@ def format_args(self, **kwargs: Any) -> str:
...
+                try:
+                    sig = inspect.signature(init, bound_method=True)
+                except TypeError:
+                    pass
+
+        # if none of the attributes are user-defined, then it's unlikely that
+        # `autodoc-before-process-signature` will be able to do anything.
+        # Let's just let inspect handle it to ensure we at least get a signature.
+        if sig is None:
+            try:
```
Comment (eric-wieser) [inline sphinx/ext/autodoc/__init__.py:1297]:
```suggestion
            # we don't know the exact method that inspect.signature will read the signature from,
            # so just pass our hook the object itself.
            self.env.app.emit('autodoc-before-process-signature', obj, False)
            try:
```

Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py:1297]: @tk0miya: is this an acceptable compromise?

Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py:1297]: @tk0miya, ping

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py:1297]: I don't understand what the comment says. And `inspect.signature()` does not refer the docstring, AFAIK. Anyway, we always pass the object itself even if we don't know the exact location.

Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py:1297]: You're right, that wording was bad. Updated.

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py:1297]: Thanks, Good to me!

**Review by tk0miya — APPROVED** [review]
LGTM

**Comment by eric-wieser** [comment]
Needs a rebase to pass tests it seems

**Comment by eric-wieser** [comment]
Rebased, hopefully CI will give me some hints this time.

File: sphinx/ext/autodoc/__init__.py [hunk 11]
```diff
@@ -1198,22 +1206,72 @@ def format_args(self, **kwargs: Any) -> str:
...
+
+        # None of the attributes are user-defined, so fall back to let inspect
+        # handle it.
+        if sig is None:
+            # We don't know the exact method that inspect.signature will read
+            # the signature from, so just pass the object itself to our hook.
+            self.env.app.emit('autodoc-before-process-signature', self.object, False)
+            try:
+                sig = inspect.signature(self.object, bound_method=False)
+            except TypeError:
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py]: It seems `inspect.signature()` raises `ValueError` also. Please handle it too.
```
Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 395, in format_signature
    args = self.format_args(**kwargs)
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 1266, in format_args
    sig = inspect.signature(self.object, bound_method=False)
  File "/Users/tkomiya/work/sphinx/sphinx/util/inspect.py", line 406, in signature
    signature = inspect.signature(subject)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 3093, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2842, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2372, in _signature_from_callable
    raise ValueError(
ValueError: no signature found for builtin type <class 'sphinx.errors.SphinxError'>
```

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: A testcase for exception class is also needed:
```
$ python
Python 3.8.2 (default, Mar  2 2020, 00:44:41)
[Clang 11.0.0 (clang-1100.0.33.17)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import inspect
>>> inspect.signature(Exception)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 3093, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2842, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2372, in _signature_from_callable
    raise ValueError(
ValueError: no signature found for builtin type <class 'Exception'>
```

Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: Test added.

**Comment by eric-wieser** [comment]
CI looks like it will pass now.

File: sphinx/ext/autodoc/__init__.py [hunk 12]
```diff
@@ -1188,22 +1196,72 @@ def format_args(self, **kwargs: Any) -> str:
...
+            if inspect.is_builtin_class_method(obj, attr):
+                return None
+            attr = self.get_attr(obj, attr, None)
+            if not (inspect.ismethod(attr) or inspect.isfunction(attr)):
+                return None
+            return attr
+
+        sig = None
+
+        # this sequence is copied from inspect._signature_from_callable
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py]: I noticed this comment now. Do we need to care for its license?

File: sphinx/ext/autodoc/__init__.py [hunk 13]
```diff
@@ -1188,22 +1196,72 @@ def format_args(self, **kwargs: Any) -> str:
...
+
+            if call is not None:
+                if "{0.__module__}.{0.__qualname__}".format(call) in _METACLASS_CALL_BLACKLIST:
+                    call = None
+
+            if call is not None:
+                self.env.app.emit('autodoc-before-process-signature', call, True)
+                try:
+                    sig = inspect.signature(call, bound_method=True)
+                except ValueError:
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py]: I think we need to handle both `ValueError` and `TypeError`.

**Review by tk0miya — COMMENTED** [review]
LGTM with nits.

File: sphinx/ext/autodoc/__init__.py (hunk 13 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: Can you come up with an example that raises `TypeError`?

File: sphinx/ext/autodoc/__init__.py (hunk 12 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: The sequence of attributes to look at is copied, but not the code itself. I don't think the license is relevant.

File: sphinx/ext/autodoc/__init__.py (hunk 13 above)
Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: I got an error with this class.
```
class Foo:
    __signature__ = 1
```
```
WARNING: error while formatting arguments for example.Foo:
Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 395, in format_signature
    args = self.format_args(**kwargs)
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 1255, in format_args
    sig = inspect.signature(self.object, bound_method=False)
  File "/Users/tkomiya/work/sphinx/sphinx/util/inspect.py", line 406, in signature
    signature = inspect.signature(subject)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 3093, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2842, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2252, in _signature_from_callable
    raise TypeError(
TypeError: unexpected object 1 in __signature__ attribute

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 398, in format_signature
    args = self.format_args()
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 1255, in format_args
    sig = inspect.signature(self.object, bound_method=False)
  File "/Users/tkomiya/work/sphinx/sphinx/util/inspect.py", line 406, in signature
    signature = inspect.signature(subject)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 3093, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2842, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2252, in _signature_from_callable
    raise TypeError(
TypeError: unexpected object 1 in __signature__ attribute
```

I know this is crazy example. But I received such report sometimes. So it would be better to handle `ValueErorr` also.

File: sphinx/ext/autodoc/__init__.py (hunk 12 above)
Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: I understand. np.

File: sphinx/ext/autodoc/__init__.py (hunk 13 above)
Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: I found related problem in autofunction: https://github.com/sphinx-doc/sphinx/issues/7629

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: I found a real bug report: https://github.com/sphinx-doc/sphinx/issues/7138

Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: That real bug report is `ValueError`, which is already handled.

I think I should try to make that traceback half as long though, pushing an update now.

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: Oh, sorry. The last one is my wrong.

File: sphinx/ext/autodoc/__init__.py [hunk 14]
```diff
@@ -391,11 +391,15 @@ def format_signature(self, **kwargs: Any) -> str:
...
-                    args = self.format_args()
+                if kwargs:
+                    try:
+                        args = self.format_args(**kwargs)
+                    except TypeError:
+                        # avoid chaining tracebacks, by putting nothing here
+                        pass
+
+                # retry without arguments for old documenters
+                args = self.format_args()
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py]: It seems `format_args()` is called twice.

Comment (eric-wieser) [inline sphinx/ext/autodoc/__init__.py]: It was called twice before too - this way, we at least:
* Do not call it twice with identical arguments
* Do not chain the exception of the first call with the exception of the second

Does anything actually pass non-empty `kwargs`?

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: >It was called twice before too

It's an unexpected result. The `TypeError` handler is added only for old Documener classes which does not accept keyword arguments. It implicitly expects `format_args()` method does not raise `TypeError`.

In this case, I think no reason to raise `TypeError` on `format_args()`. It should be handled inside it.

>Does anything actually pass non-empty kwargs?

Yes. The autosummary extension gives keyword argument.
https://github.com/sphinx-doc/sphinx/blob/62711bc081744cecba2af799ad8dbf678f6a97a4/sphinx/ext/autosummary/__init__.py#L345



Note: It's after 2 AM in Japan. So this is my last comment today. Good night!

Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: I take it back, you're correct. Updated in my latest push.

**Comment by eric-wieser** [comment]
@tk0miya: Think you'll have time to look at this again before 3.1.0?

**Comment by tk0miya** [comment]
Of course, Yes. But my request is not changed. Please catch a `TypeError` in `format_args()`. Then I'll merge this soon.

**Comment by eric-wieser** [comment]
> Please catch a TypeError in format_args(). Then I'll merge this soon.

Would you mind adding a comment on the relevant lines? I've moved things around a bit now.

File: sphinx/ext/autodoc/__init__.py (hunk 13 above)
Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: >>Please catch a TypeError in format_args(). Then I'll merge this soon.
>
>Would you mind adding a comment on the relevant lines? I've moved things around a bit now.

Sorry. As commented in this thread, `inspect.signature()` raises `TypeError`. And I think `format_args()` should not pass through it. So please catch it like `FunctionDocumenter.format_args()`.

Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py]: Thanks for the reminder, updated.

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py]: Thank you for update. I'll review this tomorrow (or in this weekend) with cleared eyes.

**Review by tk0miya — APPROVED** [review]
LGTM! Thank you for your patience and cooperation!

**Comment by eric-wieser** [comment]
I assume now #7613 and your todo comment from #3610 will be marginally easier.

**Comment by tk0miya** [comment]
Yes, I can improve #7716 now. I'm waiting for merging this :-)
